### ISSUE TITLE
Per-token host memory leak (~0.15 MB/generated token), backend-independent — reproduces on CPU, RADV, and lavapipe

### ISSUE THREAD

**Issue by novmar** [ggml-org/llama.cpp#25740]
### Name and Version

```
version: 9925 (ed8c261)
built with GNU 15.2.0 for Linux x86_64
```
Backend: Vulkan (RADV, Mesa 26.1.4). Also reproduced on **lavapipe** (see below).

### Operating systems

Linux

### Which llama.cpp modules do you know to be affected?

llama-server

### Command line

```shell
llama-server -m qwen3-14b-q4km.gguf -ngl 99 -c 8192 -b 512 -ub 512 -fa on --no-mmap -t 6 --host 0.0.0.0 --port 8080
```

### Problem description & steps to reproduce

`llama-server` on the **Vulkan** backend leaks host memory that scales with the number of **generated tokens**, not with the number of requests. The process RSS grows linearly and is only reclaimed when the process exits; under sustained generation it eventually exhausts system RAM.

I measured this carefully and isolated it. Hardware: AMD BC-250 (PS5 "Ariel" APU, GPU **Cyan Skillfish / gfx1013**, RDNA2). Model: Qwen3-14B Q4_K_M.

**1. It is a pure host-RSS leak.** During sustained `/completion` load, `llama-server` RSS grows linearly while **GPU/GTT memory, VRAM, kernel slab (SUnreclaim/SReclaimable) and PageTables all stay flat**. `MemAvailable` declines 1:1 with RSS. Killing the process returns *all* of it (avail 4.8 GB → 14.5 GB, GTT 9.6 GB → 12 MB).

**2. It scales per generated token (~0.16 MB/token), not per request.**

| request `n_predict` | RSS growth per request |
|---|---|
| 8 tokens | ~0 (flat, ~1 MB/req — noise) |
| 400 tokens | ~57–64 MB/req |

64 MB / 400 ≈ **0.16 MB per generated token**. So it appears to be a per-graph-submission (per-decode) allocation that is never freed.

**3. It is NOT flash attention.** Identical slope with `-fa on` (57 MB/req) vs `-fa off` (59 MB/req) at `n_predict=400`. `-fa off` only lowers the baseline RSS by ~50 MB; the leak rate is unchanged.

**4. It is NOT the driver (RADV) — it is in the ggml-vulkan backend.** I ran the **exact same binary** through **lavapipe** (`VK_ICD_FILENAMES=…/lvp_icd.x86_64.json`) instead of RADV. The leak persists at essentially the same per-token rate:

| driver | leak |
|---|---|
| RADV (Mesa 26.1.4) | ~0.16 MB/token |
| lavapipe (llvmpipe) | ~0.12 MB/token (47 MB over 384 tokens) |

Since a completely different Vulkan ICD leaks at the same per-token rate, the allocation that leaks lives in the **ggml-vulkan backend code**, not in the AMD/RADV driver. (This also means a Mesa upgrade does not help.)

This is on the latest build (b9925) and latest released Mesa (26.1.4). It looks related to the older, still-open report #12531 (Vulkan per-request growth on RX 6600 XT), but here it is quantified per-token and shown to be driver-independent.

### Reproduction script (sketch)

```sh
# start server (Vulkan), then:
for i in $(seq 1 15); do
  curl -s http://127.0.0.1:8080/completion -d '{"prompt":"...","n_predict":400}' >/dev/null
  ps -o rss= -C llama-server   # grows ~60 MB each iteration; flat if n_predict=8
done
```

Happy to run further diagnostics (specific ggml-vulkan ops, valgrind/massif on lavapipe, different models) — I have a reliable per-token measurement harness set up.

**Comment by novmar** [comment]
**Update / correction: this is NOT Vulkan-specific — it reproduces on the pure CPU backend.**

I re-ran the exact per-token RSS measurement with the GPU/Vulkan completely out of the picture (Vulkan disabled via empty ICD, `-ngl 0`, confirmed CPU-only: GTT ≈ 12 MB, model resident in host RSS at ~9.3 GB baseline). Same build b9925 (ed8c261), same model, same `llama-server`.

Result — the leak is unchanged:

| backend | leak rate |
|---|---|
| Vulkan / RADV (AMD gfx1013) | ~0.16 MB/token |
| Vulkan / lavapipe (llvmpipe) | ~0.12 MB/token |
| **CPU (Vulkan disabled, -ngl 0)** | **0.145 MB/token** (+232 MB over 1600 generated tokens) |

So the per-token host-RSS growth is **backend-independent** — it happens identically on CPU with no Vulkan involved. This means the leak is **not in the ggml-vulkan backend** (my original title was wrong); it is in **ggml core / llama.cpp server-side per-decode bookkeeping** (candidates: ggml-alloc / `ggml_backend_sched`, or llama-server slot/logits/sampler state).

Still a reachable-growth leak (freed on process exit), so ASAN/valgrind-memcheck report nothing — heaptrack/massif on the CPU backend (simplest repro, no GPU driver noise) is the way to localize it. I have a reliable per-token RSS harness and can run heaptrack if that helps.

Apologies for the misleading "Vulkan backend" framing in the title — updating it to reflect the CPU reproduction.

**Comment by novmar** [comment]
**Further isolation: the leak rate is architecture-dependent — Qwen3 models leak ~2.5× more than Qwen2.5, independent of size.**

Same machine, same `llama-server`, same pure-CPU setup (Vulkan disabled, `-ngl 0`, confirmed GTT≈0), 1600 generated tokens each:

| model | leak |
|---|---|
| Qwen2.5-7B-Q4 | **0.052 MB/token** |
| Qwen3-8B-Q4 | **0.131 MB/token** |
| Qwen3-14B-Q4 | **0.145 MB/token** |

Key point: **Qwen3-8B (0.131) ≈ Qwen3-14B (0.145)**, while the similarly-sized **Qwen2.5-7B leaks ~2.5× less (0.052)**. So the per-token host-RSS growth is **not proportional to model size** — it's tied to something in the **Qwen3 architecture / graph** that Qwen2.5 doesn't exercise (candidates: Qwen3's per-head QK-norm, or another Qwen3-specific op/allocation in the decode graph).

Combined with the earlier results — backend-independent (CPU, RADV, lavapipe all identical), version-independent (b9842 == b9925), not flash-attention, reachable-growth (freed on exit) — this points at a per-decode host allocation on a **Qwen3-specific code path** in ggml-core / llama.cpp, not in any backend.

For anyone reproducing: use a Qwen3 model (the leak is ~2.5× more visible than with Qwen2.5) and a growth profiler (heaptrack/massif) — not ASAN/valgrind-memcheck, which show nothing since the memory is reachable and freed at exit.

**Comment by novmar** [comment]
**Correction to my previous comment: the leak is NOT Qwen3-specific — it's universal and scales with model dimensions.** Testing a non-Qwen architecture (Llama-3.1-8B) settles it.

Same machine, pure CPU (`-ngl 0`, Vulkan disabled), 1600 generated tokens each:

| model | hidden size | leak |
|---|---|---|
| Qwen2.5-7B | 3584 | 0.052 MB/token |
| **Llama-3.1-8B** | 4096 | **0.116 MB/token** |
| Qwen3-8B | 4096 | 0.131 MB/token |
| Qwen3-14B | 5120 | 0.145 MB/token |

At the **same hidden size (4096), Llama-3.1 and Qwen3-8B leak essentially the same** (0.116 vs 0.131) — so this is **not** a Qwen3 architecture feature (I was wrong to suggest QK-norm). The leak grows monotonically with model dimensions across architectures; Qwen2.5-7B leaked least simply because it's the smallest (hidden 3584).

So the corrected picture: a **per-decode host allocation whose size is proportional to the model's hidden/KV dimensions**, leaked once per generated token, on every backend (CPU/RADV/lavapipe) and every model family tested. That shape strongly suggests a per-token buffer sized like the KV-cache slice or a hidden-state/embedding tensor that is allocated each decode and never freed/reused (reachable until process exit).

Summary of what's been ruled out: backend (CPU==RADV==lavapipe), llama.cpp version (b9842==b9925), flash-attention (on==off), and model architecture (Llama==Qwen3 at equal size). Remaining: a size-proportional per-decode host buffer in ggml-core / llama.cpp. heaptrack/massif on the CPU backend is the next step to name it.

**Comment by novmar** [comment]
**Resolved — this is not a leak. It's the `llama-server` host-RAM prompt cache (`--cache-ram`, default 8192 MiB), behaving as designed (PR #16391).**

Full root-cause with heaptrack (RelWithDebInfo, fully symbolicated, on b9925+1):

- **Bare decode does NOT grow.** `llama-cli -ngl 0 -n 500` on CPU grew ~480 KB total (~1 KB/token). A *single* `llama-server` request is also flat during generation. So there is no per-token leak in the decode/graph path — my earlier measurements were all on `llama-server` across *repeated* requests.
- **The growth is per finished request, not per token.** 10 sequential `/completion` requests (distinct prompts) grew RSS **+17,068 KiB per request, perfectly linear (10/10)**. 17,068 KiB ÷ 118 tokens = **144.6 KiB/token ≈ the exact KV-state size for Qwen3-8B** (2 × 36 layers × 8 KV-heads × 128 × 2 B = 144.0 KiB/token). The earlier cross-model scaling (0.052 / 0.116 / 0.131 / 0.145 MB/tok for Qwen2.5-7B / Llama-3.1-8B / Qwen3-8B / Qwen3-14B) is ~0.89× each model's KV-bytes/token — it was always the KV state being copied to host.
- **Allocation site** (heaptrack, dominant growing backtrace):
  ```
  std::vector<uint8_t>::resize
    → server_prompt_cache::alloc      tools/server/server-task.cpp:1709
    → server_slot::prompt_save        tools/server/server-context.cpp:231/236  (llama_state_seq_get_data_ext)
    → server_context_impl::get_available_slot  server-context.cpp:1620
    → process_single_task / start_loop
  ```
  On slot reuse (and idle-slot save, `cache_idle_slots` default true), the slot's entire KV state is serialized into a host `std::vector` in `server_prompt_cache::states`, LRU-evicted only above `cache_ram_mib` (default **8192**, `common/common.h:627`; `cache_prompt` default true).

This explains every property that made it look like a leak: host-only, backend-independent (state is always copied to host, so identical on CPU / RADV / lavapipe), scales with KV width, survives across requests, freed only at exit (fully reachable → invisible to ASAN/valgrind). On a memory-constrained box it grows toward the 8 GiB cap and can OOM the machine before the cap is reached.

**Fix (verified): `--cache-ram 0`** (or a sane limit). Before: 17,068 KiB/request. After: RSS flat at 127 MB across 8+ requests (steady-state +72 KiB/request). Confirmed on the original hardware.

The one thing that arguably *is* worth improving upstream: the prompt-cache config messages are all `SRV_TRC` (server-context.cpp:1339–1352), so at default verbosity the server never mentions that it's holding up to 8 GiB of KV snapshots — which is exactly why this reads as a leak. Promoting those to `SRV_INF` (a one-liner like "prompt cache enabled, limit 8192 MiB, use --cache-ram 0 to disable") would prevent others from chasing this. Happy to open a separate small PR for that if useful.

Apologies for the mis-filed "leak" framing — closing this as not-a-bug. Thanks to the heaptrack pass for pinning it exactly.
